README.md
<p align="center">
  <a href="#" target="blank"><img src="./.github/docs/media/studysnap.png" width="320" alt="StudySnap Logo" /></a>
</p>

<div align="center">

[![Status](https://img.shields.io/badge/status-active-success.svg)]()
![Build and Push Docker](https://github.com/Study-Snap/authentication/workflows/Build%20and%20Push%20Docker/badge.svg)
![Lint and Scan Docker](https://github.com/Study-Snap/authentication/workflows/Lint%20and%20Scan%20Docker/badge.svg)
![E2E and Coverage](https://github.com/Study-Snap/authentication/workflows/E2E%20and%20Coverage/badge.svg)
![Unit Tests](https://github.com/Study-Snap/authentication/workflows/Unit%20Tests/badge.svg)
[![License](https://img.shields.io/badge/license-Apache2.0-blue.svg)](/LICENSE)

</div>

---

<p align="center">Authentication Backend for the StudySnap application</p>
    <p align="center">

## Description

The authentication backend for StudySnap created using the [NestJS](http://nestjs.com) framework. Leverages [Passport](http://www.passportjs.org) for use of custom authentication strategies and JWT token signing.

## Useful Links

- Check out the [Authentication Helm Chart](https://github.com/Study-Snap/charts/tree/master/charts/authentication) used for helm deployment
- This project tracks changes through the [CHANGELOG](/CHANGELOG.md)

## Prerequisites

This project has a few extra requirements in order to function properly.

- Connection to an existing(or new) PostgreSQL database (configured through docker). This is where we would like to store users and active refresh tokens.
- **Preferrably** [Docker](http://docker.com) and/or [docker-compose](https://docs.docker.com/compose/). Note: Can be run without docker.
- Configured `.env` or `exported` environment variables according to the available configurations listed below.

## Available Configurations

Below is a list of available configuration options to customize the project. **Note:** These configuration options are available to be individually configured as per your environment (`development`, `test`, and `production`)

| Option                  | Description                                                                                              | Default                   | Optional |
|-------------------------|----------------------------------------------------------------------------------------------------------|---------------------------|----------|
| PORT             | Defines the port for the API to listen on                                                                | `5555`                    | Y        |
| MAX_REQUESTS            | Defines the maximum number of requests per 15 minutes (rate limiting)                                    | `250`                    | Y        |
| BCRYPT_SALT_ROUNDS           | Specifies the number of salt rounds to apply to password hashes using [bcrypt](https://www.npmjs.com/package/bcrypt) | `12` | Y        |
| DB_DIALECT       | Specifies the type of database you wish to use in your implementation. Thanks to NestJS, this is optional, however there will be some limited extra setup required for anything other than `postgreSQL` | `postgres`                      | Y        |
| DB_HOST                | Specifies the database host address (IP or Domain) to reach the database                                | `localhost`               | N        |
| DB_PORT                | Specifies the port to reach the database host application                                                                    | `5432`                    | Y        |
| DB_USER                | The database user to authenticate to the database host                                                    | `NONE`                     | N        |
| DB_PASS            | The password to authenticate `$DB_USER`                                                             | `NONE`                     | N        |
| DB_USER_DATABASE       | The name of the application database where you will store your users.                                                 | `studysnap_db`               | Y        |
| DB_RETRY_ATTEMPTS             | Number of times to retry a failed connection to the database configured.                                                                              | `2`                   | Y        |
| JWT_SECRET      | Used to cryptographically sign/decode authentication (JWT) tokens sent/recieved from the authentication/authorization server.                                                                              | `NONE`                     | N        |
| JWT_ACCESS_TOKEN_EXPIRE_TIME | Specifies the time it takes for an access token to expire. (in `production` it is recommended this be a short-lived token to limit risk of compromised tokens causing damange) | `10m`                       | Y        |

> The current dev environment setup I have convieniently included in the `.env` file at the root of this project.

## Running the app

You can choose to run this project in multiple ways depending on what you need.

### Extra Steps

Currently there is one additional setup command you will need to run before running the app in Docker or Standalone. You must have postgreSQL installed and configured properly (as per **prerequisites** above). Do this with Docker (**recommended**).

```bash

# Usable with default configuration
$ docker run -d \
    -e POSTGRES_DB=studysnap_db \
    -e POSTGRES_USER=studysnap \
    -e POSTGRES_PASSWORD=snapstudy \
    -p 5432:5432 \
    postgres:13.1

```

> Make sure that you change the passed environment variables if your needs change.

### Docker (recommended)

From the project root, run the following.

```bash
# Run from published image
$ docker run -d -p 5555:5555 <other_options> studysnap/authentication:<version_tag>

# Build & Run Locally
$ docker build -t local/studysnap-authentication:latest .
$ docker run -d -p 5555:5555 <other_options> local/studysnap-authentication:latest

```

### Standalone / Development

Install the required project dependencies

```bash
$ npm install
```

Run the project

```bash
# development
$ npm run start

# development with watch mode (recommended)
$ npm run start:dev

# production mode
$ npm run start:prod
```

## Running Tests

First, for e2e testing, make sure that you have a test database running in the environment.

```bash

# Use docker to start up a postgres database for testing
docker run -d -e POSTGRES_DB=studysnap_testdb -e POSTGRES_USER=studysnap -e POSTGRES_PASSWORD=snapstudy -p 7654:5432 postgres:13.1

```

Now you can run any of the following to test your changes

```bash
# unit tests
$ npm run test

# e2e tests
$ npm run test:e2e

# test coverage
$ npm run test:cov
```

## Deployment

We are currently using Helm to handle our deployment to our K8s Cluster. To manually deploy, we need to execute the following commands.

```bash

# Add the studysnap helm repository & Update
helm repo add studysnap https://study-snap.github.io/charts/
helm repo update

# Now install the authentication chart from the repository
helm install authentication studysnap/authentication [optional_flags]

```

> I personally recommend using [k8s Lens](http://k8slens.dev) to help visualize the deployment on the cluster

## Support

Create an **issue** in the StudySnap [Jira](http://studysnap.atlassian.net)

## Authors

- [Benjamin Sykes](https://sykesdev.ca)
- [Liam Stickney](https://github.com/LiamStickney)
- [Malik Sheharyaar Talhat](https://github.com/maliksheharyaar)

## License

StudySnap is [Apache licensed](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Released]

## [1.0.4]

### Added

- (SSPP-20): User password change feature exposed to API
- (dev): Added development ready docker-compose configuration to support future development of Authentication for StudySnap
- (SSPP-20): Implemented some more e2e tests
- (SSPP-20): Some functional support services for email modification
- (SSPP-302): Added some extra documentation to classes and functions for better usability for devs
- (SSPP-114): Implemented logging of all authentication requests from users (registration and validation of user credentials as well as updates to user data)

## [1.0.3]

### Added

- (SSPP-248): CI/CD tagging with unique commit hash for testing on bleeding-edge releases

## [1.0.2]

### Added

- (REF-PATHS): Added test docker-compose to simply dev testing
- (REF-PATHS): Added test docker-push to workflows for testing branched releases without waiting for staging

### Modified

- (REF-PATHS): Fixed pathing to match new DNS scheme
- (REF-PATHS): Updated test-e2e workflow to include environment setup with docker-compose
- (SSPP-192): Updated environment to match new classrooms architecture

## [1.0.1]

### Modified

- (hotfix): Fixed pathing for controller endpoints so that k8s deployment does not conflict with studysnap/neptune deployment.
- (ref): Refactored some code regarding `/auth/docs`
- (hotfix): Hopefully resolve any further issues with environment confusion by removing (renaming) `.env` to a development variant and excluding further `*.env` files from git and docker builds.

## [1.0.0]

### Added

- (SSPP-69): Swagger UI implementation available on `domain.com/docs` to help support our program documentation.
- (SSPP-77): Support for encrypted connection through SSL
- (docs): Added changelog link to main README as well as connecting link for helm chart
- (SSPP-90): CI/CD to support postgres server package in helm chart
- (SSPP-90): Octopus Deployment release process in CD

### Modified

- (SSPP-25): Modified token refresh endpoint to accept refreshToken in the form of an httpOnly cookie instead of body request.
- (SSPP-25): Updated e2e testing suite to set cookie instead for testing refresh token functionality
- (docs): Updated docs with deploy instructions
- (docs): Fixed badges on docs
- (SSPP-69): Refactored some source code

### Removed

- (SSPP-25): The old DTO for tokenRefresh as we do not require it anymore since it is impossible to pass a refreshToken using the request body

## [0.1.0]

### Added

- (docs-init): Documentation for requirements, setup and deployment instructions
- (auto-cicd): Github workflows to help automate lint, test, build and publishing of this application
- (local-auth): Local authentication that generates JWT Access & Refresh tokens for stateless authentication
- (auto-test): Unit and End-to-End (e2e) testing captured in CI process.
- (docs): Instructions for running without docker in development

### Modified

- (refactored): Codebase refactored to be easier to navigate

### Removed

- (auto-cicd): Some annoying extra linting rules provided in the recommended set.


src/config/index.ts
import { DEVELOPMENT, PRODUCTION, TEST } from '../common/constants'
import { IConfig } from '../common/interfaces/config/app-config.interface'
import * as dotenv from 'dotenv'

// Import env vars from appropriate file
switch (process.env.NODE_ENV) {
	case DEVELOPMENT:
		dotenv.config({ path: '.dev.env' })
		break
	case TEST:
		dotenv.config({ path: '.test.env' })
	case PRODUCTION:
		dotenv.config({ path: '.prod.env' })
	default:
		dotenv.config()
		break
}

const appConfig: IConfig = {
	development: {
		listenPort: process.env.PORT || 5555,
		maxRequests: parseInt(process.env.MAX_REQUESTS) || 250,
		bcryptSaltRounds: parseInt(process.env.BCRYPT_SALT_ROUNDS) || 12,
		dbDialect: process.env.DB_DIALECT || 'postgres',
		dbHost: process.env.DB_HOST || 'localhost',
		dbPort: process.env.DB_PORT || 5432,
		dbUsername: process.env.DB_USER,
		dbPassword: process.env.DB_PASS,
		dbDatabaseName: process.env.DB_DATABASE_NAME || 'studysnap_db',
		dbRetryAttempts: parseInt(process.env.DB_RETRY_ATTEMPTS) || 2,
		jwtSecret: process.env.JWT_SECRET || 'dev_secret_do_change_in_prod',
		jwtAccessExpireTime: process.env.JWT_ACCESS_TOKEN_EXPIRE_TIME || '20m'
	},
	test: {
		listenPort: process.env.PORT || 5000,
		maxRequests: 999,
		bcryptSaltRounds: parseInt(process.env.BCRYPT_SALT_ROUNDS) || 12,
		dbDialect: process.env.DB_DIALECT || 'postgres',
		dbHost: process.env.DB_HOST || 'localhost',
		dbPort: process.env.DB_PORT || 7654,
		dbUsername: process.env.DB_USER || 'studysnap',
		dbPassword: process.env.DB_PASS || 'snapstudy',
		dbDatabaseName: process.env.DB_DATABASE_NAME || 'studysnap_db',
		dbRetryAttempts: parseInt(process.env.DB_RETRY_ATTEMPTS) || 2,
		jwtSecret: process.env.JWT_SECRET || 'test_secret',
		jwtAccessExpireTime: '10h'
	},
	production: {
		listenPort: process.env.PORT || 5555,
		maxRequests: parseInt(process.env.MAX_REQUESTS) || 250,
		bcryptSaltRounds: parseInt(process.env.BCRYPT_SALT_ROUNDS) || 12,
		dbDialect: process.env.DB_DIALECT || 'postgres',
		dbHost: process.env.DB_HOST,
		dbPort: process.env.DB_PORT || 5432,
		dbUsername: process.env.DB_USER,
		dbPassword: process.env.DB_PASS,
		dbDatabaseName: process.env.DB_DATABASE_NAME || 'studysnap_db',
		dbRetryAttempts: parseInt(process.env.DB_RETRY_ATTEMPTS) || 5,
		jwtSecret: process.env.JWT_SECRET,
		jwtAccessExpireTime: process.env.JWT_ACCESS_TOKEN_EXPIRE_TIME || '10m'
	}
}

/**
 * Gets appropriate configuration for execution enviornment in any area of the app the includes this function.
 * @returns Appropriate configuration for the execution environment
 */
export const getConfig = () => {
	switch (process.env.NODE_ENV) {
		case DEVELOPMENT:
			return appConfig.development
		case TEST:
			return appConfig.test
		case PRODUCTION:
			return appConfig.production
		default:
			return appConfig.development
	}
}
